sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImplTest.java

```diff
@@ -126,6 +149,313 @@ public void testRead() throws IOException {
     verifyMetricWasSet("google.bigtable.v2.ReadRows", "ok", 1);
   }
 
+  /**
+   * This test ensures that protobuf creation and interactions with {@link BigtableDataClient} work
+   * as expected. This test checks that a single row is returned from the future.
+   *
+   * @throws IOException
+   */
+  @Test
+  public void testReadSingleRangeBelowSegmentLimit() throws Exception {
+    ByteKey start = ByteKey.copyFrom("a".getBytes(StandardCharsets.UTF_8));
+    ByteKey end = ByteKey.copyFrom("b".getBytes(StandardCharsets.UTF_8));
+
+    when(mockBigtableSource.getRanges()).thenReturn(Arrays.asList(ByteKeyRange.of(start, end)));
+    FlatRow expectedRow = FlatRow.newBuilder().withRowKey(ByteString.copyFromUtf8("a")).build();
+
+    when(mockBigtableDataClient.readFlatRows(any(ReadRowsRequest.class), any()))
+        .thenAnswer(mockReadRowsAnswer(Arrays.asList(expectedRow)));
+
+    BigtableService.Reader underTest =
+        BigtableServiceImpl.BigtableSegmentReaderImpl.create(mockSession, mockBigtableSource);
+
+    underTest.start();
+    Assert.assertEquals(FlatRowConverter.convert(expectedRow), underTest.getCurrentRow());
+    Assert.assertFalse(underTest.advance());
+    underTest.close();
+
+    verifyMetricWasSet("google.bigtable.v2.ReadRows", "ok", 2);
+  }
+
+  /**
+   * This test ensures that all the rows are properly added to the buffer and read. This example
+   * uses a single range with SEGMENT_SIZE*2+1 rows. Range: [a, b1, ... b199, c)
+   *
+   * @throws IOException
+   */
+  @Test
+  public void testReadSingleRangeAboveSegmentLimit() throws IOException {
+    ByteKey start = ByteKey.copyFrom("b00000".getBytes(StandardCharsets.UTF_8));
+    ByteKey end = ByteKey.copyFrom(("b00"+SEGMENT_SIZE*2).getBytes(StandardCharsets.UTF_8));
+
+    when(mockBigtableSource.getRanges()).thenReturn(Arrays.asList(ByteKeyRange.of(start, end)));
+
+    List<List<FlatRow>> expectedResults = ImmutableList.of(
+        generateSegmentResult("b",0,SEGMENT_SIZE, false),
```

Comment:
Can we simplify this to generateSegmentResult("b", 0, SEGMENT_SIZE * 2, false)?

Reply:
The reason for the 2D list is to allow for stub.thenAnswer to return a full List<Row> instead of having to sublist up a larger list differently in each test.